lib/SILGen/SILGenExpr.cpp

```diff
@@ -3337,28 +3338,67 @@ lowerKeyPathSubscriptIndexTypes(
     indexLoweredTy = indexLoweredTy.mapTypeOutOfContext();
     indexPatterns.push_back({indexTy->mapTypeOutOfContext()
                                     ->getCanonicalType(),
-                             indexLoweredTy});
+                             indexLoweredTy, index->isVariadic()});
   }
 };
 
-static void
-lowerKeyPathSubscriptIndexPatterns(
-                 SmallVectorImpl<KeyPathPatternComponent::Index> &indexPatterns,
-                 ArrayRef<IndexTypePair> indexTypes,
-                 ArrayRef<ProtocolConformanceRef> indexHashables,
-                 unsigned &baseOperand) {
+static void lowerKeyPathSubscriptEqualsIndexPatterns(
+    SmallVectorImpl<KeyPathPatternComponent::Index> &indexPatterns,
+    ArrayRef<IndexTypeTuple> indexTypes,
+    ArrayRef<ProtocolConformanceRef> indexHashables, unsigned baseOperand,
+    SILGenModule &SGM, ResilienceExpansion expansion) {
   for (unsigned i : indices(indexTypes)) {
     CanType formalTy;
     SILType loweredTy;
-    std::tie(formalTy, loweredTy) = indexTypes[i];
+    bool isVariadic;
+    std::tie(formalTy, loweredTy, isVariadic) = indexTypes[i];
     auto hashable = indexHashables[i].mapConformanceOutOfContext();
+    // We have an array of variadic parameters...
+    if (isVariadic) {
+      // Use the element type instead of the array type for hash/equals.
+      auto arrayTy = cast<BoundGenericStructType>(formalTy.getPointer());
+      auto elementTy = arrayTy->getGenericArgs()[0];
```

Comment:
I'm not sure we want to be using the element type here, given the variadic argument is passed to the key path as an array, so using the array's implementation of Hashable/Equatable should be correct. 

From what I can tell, the problem appears to be that CSApply is retrieving the Hashable conformance of the element instead of the array for a variadic param. I believe it should be possible to change `getPlainType` to `getParameterType` here to fix this:

https://github.com/apple/swift/blob/d2e1f0916976583e6bba20cf550b9d9cfdb62612/lib/Sema/CSApply.cpp#L4594-L4601

For symmetry we should probably also change it for the solver logic too:

https://github.com/apple/swift/blob/d2e1f0916976583e6bba20cf550b9d9cfdb62612/lib/Sema/ConstraintSystem.cpp#L2253

Reply:
You're right. That's a much better and cleaner fix.